### PR TITLE
mqtt: fix mqtt msgid err

### PR DESCRIPTION
**Pull request by Txiaozhe**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by lolBig** [comment]
please provide documents for subscribe header and describe why dose `this._msgId++` not working

**Comment by Txiaozhe** [comment]
The control packet of SUBSCRIBE, UNSUBSCRIBE and PUBLISH(QoS>0) must contains a nonzero packet identifier(look at [doc 2.3.1](https://mcxiaoke.gitbooks.io/mqtt-cn/content/mqtt/02-ControlPacketFormat.html))
our code(e.e.: subscribe ):
```javascript
var buf = this._handle._getSubscribe(topic, {
      id:  this._msgId++,
      qos: (options && options.qos) || 0,
});
```
The value of `this._msgId` will be `0` if we use `this._msgId++` to generate the packet identifier and `++this._msgId` is needed.

**Comment by lolBig** [comment]
LGTM

**Comment by yorkie** [comment]
> The control packet of SUBSCRIBE, UNSUBSCRIBE and PUBLISH(QoS>0) must contains a nonzero packet identifier(look at [doc 2.3.1](https://mcxiaoke.gitbooks.io/mqtt-cn/content/mqtt/02-ControlPacketFormat.html))
> our code(e.e.: subscribe ):
> 
> ```js
> var buf = this._handle._getSubscribe(topic, {
>       id:  this._msgId++,
>       qos: (options && options.qos) || 0,
> });
> ```
> 
> The value of `this._msgId` will be `0` if we use `this._msgId++` to generate the packet identifier and `++this._msgId` is needed.

Could you please reference the official documentation?

**Comment by Txiaozhe** [comment]
Done, could please review it?

File: src/js/mqtt.js [hunk 1]
```diff
@@ -274,6 +274,11 @@ MqttClient.prototype.disconnect = function(err) {
   this._socket.end();
 };
 
+function qosEqual0(qos) {
+  if (!qos || Number.isNaN(qos)) return false;
```
Comment (yorkie) [inline src/js/mqtt.js]: What the case that qos is NaN?

File: src/js/mqtt.js [hunk 2]
```diff
@@ -274,6 +274,11 @@ MqttClient.prototype.disconnect = function(err) {
...
+  return qos === 0;
+}
```
Comment (yorkie) [inline src/js/mqtt.js]: And I'd like to add a `normalizeOpts()` function to replace this, we should use `(options && options.qos) || 0`.

File: src/js/mqtt.js [hunk 3]
```diff
@@ -340,7 +349,7 @@ MqttClient.prototype.unsubscribe = function(topic, callback) {
   // TODO don't use try catch
   try {
     buf = this._handle._getUnsubscribe(topic, {
-      id: this._msgId++,
+      id: ++this._msgId,
```
Comment (yorkie) [inline src/js/mqtt.js]: QoS is missing here?

**Comment by yorkie** [comment]
> Done, could please review it?

Where is the official spec to describe this? If those are protocol-specified feature, it's better to change at MQTTPacket.

File: test/run_pass/test_mqtt_sub_pub.js [hunk 4]
```diff
@@ -28,7 +28,9 @@ Promise.all([
   babeee.subscribe('u/love', common.mustCall(function() {
     console.log('babeee subscribed u/love');
     setTimeout(function() {
-      yorkie.publish('u/love', 'endless');
+      yorkie.publish('u/love', 'endless', common.mustCall(function() {
+        console.log('yorkie published u/love');
+      }));
```
Comment (yorkie) [inline test/run_pass/test_mqtt_sub_pub.js]: What do you wanna test by this change?

File: src/js/mqtt.js (hunk 1 above)
Comment (Txiaozhe) [reply src/js/mqtt.js]: when the `options` argument is undefined

**Comment by yorkie** [comment]
Note that adding a related label is also required if the PR is opened by collaborators.

File: src/js/mqtt.js (hunk 3 above)
Comment (Txiaozhe) [reply src/js/mqtt.js]: There is not `options` argument here.

File: src/js/mqtt.js [hunk 5]
```diff
@@ -283,12 +288,16 @@ MqttClient.prototype.disconnect = function(err) {
  */
 MqttClient.prototype.publish = function(topic, payload, options, callback) {
   callback = callback || noop;
+  if (typeof options === 'function') {
+    callback = options;
+  }
```
Comment (yorkie) [inline src/js/mqtt.js]: This is unrelated, and it changes the API, revert this, please.

File: test/run_pass/test_mqtt_sub_pub.js (hunk 4 above)
Comment (Txiaozhe) [reply test/run_pass/test_mqtt_sub_pub.js]: I check the case when options is undefined and the callback is the third argument, may it is not very rigorous, I will revise it.

File: src/js/mqtt.js (hunk 5 above)
Comment (Txiaozhe) [reply src/js/mqtt.js]: Okay.

File: src/js/mqtt.js [hunk 6]
```diff
@@ -283,12 +288,16 @@ MqttClient.prototype.disconnect = function(err) {
...
+
   if (!Buffer.isBuffer(payload)) {
     payload = new Buffer(payload);
   }
   try {
     var buf = this._handle._getPublish(topic, {
-      id: this._msgId++,
+      id: qosEqual0(options && options.qos) ? 0 : ++this._msgId,
```
Comment (yorkie) [inline src/js/mqtt.js]: ++ is not good way to represent, the root reason is that the `msgId` is starting from zero, could that start from 1 to fix this problem?

Comment (Txiaozhe) [reply src/js/mqtt.js]: But should it be the different value when different publish?

Comment (Txiaozhe) [reply src/js/mqtt.js]: Yeah, may ++ is not very readable:)

File: src/js/mqtt.js (hunk 1 above)
Comment (yorkie) [reply src/js/mqtt.js]: undefined has been covered at `!qos`

File: src/js/mqtt.js (hunk 3 above)
Comment (yorkie) [reply src/js/mqtt.js]: > There is not `options` argument here.

What I mean is that if the unsubscribe need qos option?

File: src/js/mqtt.js [hunk 7]
```diff
@@ -317,7 +323,7 @@ MqttClient.prototype.subscribe = function(topic, options, callback) {
   }
   try {
     var buf = this._handle._getSubscribe(topic, {
-      id: this._msgId++,
+      id: qosEqual0(options.qos) ? 0 : this._msgId++,
       qos: (options && options.qos) || 0,
```
Comment (yorkie) [inline src/js/mqtt.js]: Subscribe's qos could be removed from this option, see https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901023, only the publish's qos is used at MQTT v5.0-spec.

Comment (Txiaozhe) [reply src/js/mqtt.js]: But our mqtt version is 3.x now, [http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718063](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718063) 
> The payload of a SUBSCRIBE packet MUST contain at least one Topic Filter / QoS pair. A SUBSCRIBE packet with no payload is a protocol violation

**Comment by yorkie** [comment]
@Txiaozhe What you want to fix is documented here: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901026.

> These MQTT Control Packets are PUBLISH (where QoS > 0), PUBACK, PUBREC, PUBREL, PUBCOMP, SUBSCRIBE, SUBACK, UNSUBSCRIBE, UNSUBACK.

As the above description, I think we only check QoS at the `.publish()`.

> Each time a Client sends a new SUBSCRIBE, UNSUBSCRIBE,or PUBLISH (where QoS > 0) MQTT Control Packet it MUST assign it a non-zero Packet Identifier that is currently unused

Firstly we should make the packet id namely `msgId` to be a non-zero Two Byte Integer, and is currently unused. Otherwise we only need do SUBSCRIBE, UNSUBSCRIBE,or PUBLISH (where QoS > 0) MQTT.

File: src/js/mqtt.js (hunk 3 above)
Comment (Txiaozhe) [reply src/js/mqtt.js]: there is not `qos` option in unsubscribe.

File: src/js/mqtt.js [hunk 8]
```diff
@@ -274,6 +276,19 @@ MqttClient.prototype.disconnect = function(err) {
   this._socket.end();
 };
 
+function qosEqual0(qos) {
```
Comment (legendecas) [inline src/js/mqtt.js]: `qosEqual0` would still return `false` if `qos` is `0`.

File: src/js/mqtt.js [hunk 9]
```diff
@@ -22,6 +22,8 @@ var MQTT_PINGRESP = 13;
 var MQTT_DISCONNECT = 14;
 /* eslint-enable */
 
+var MAX_MSG_ID = 65535;
```
Comment (yorkie) [inline src/js/mqtt.js]:
```suggestion
var MAX_MSG_ID = 0xffff;
```

File: src/js/mqtt.js [hunk 10]
```diff
@@ -274,6 +276,18 @@ MqttClient.prototype.disconnect = function(err) {
   this._socket.end();
 };
 
+function getQoS(qos) {
+  return Number.isNaN(qos) ? 0 : (qos >= 0 && qos <= 2 ? qos : 0);
+}
+
+MqttClient.prototype._nextMsgId = function() {
+  if (this._msgId > MAX_MSG_ID) {
+    this._msgId = 1;
```
Comment (yorkie) [inline src/js/mqtt.js]: The spec shows the packet id won't be used, that implies us to store the range of identifiers used?

File: src/js/mqtt.js [hunk 11]
```diff
@@ -274,6 +276,18 @@ MqttClient.prototype.disconnect = function(err) {
   this._socket.end();
 };
 
+function getQoS(qos) {
+  return Number.isNaN(qos) ? 0 : (qos >= 0 && qos <= 2 ? qos : 0);
+}
+
+MqttClient.prototype._nextMsgId = function() {
+  if (this._msgId > MAX_MSG_ID) {
```
Comment (yorkie) [inline src/js/mqtt.js]: How about renaming as `_packetId`?

File: src/js/mqtt.js [hunk 12]
```diff
@@ -274,6 +276,18 @@ MqttClient.prototype.disconnect = function(err) {
   this._socket.end();
 };
 
+function getQoS(qos) {
+  return Number.isNaN(qos) ? 0 : (qos >= 0 && qos <= 2 ? qos : 0);
+}
+
+MqttClient.prototype._nextMsgId = function() {
```
Comment (yorkie) [inline src/js/mqtt.js]: Suggestion: getNewPacketId()

File: src/js/mqtt.js [hunk 13]
```diff
@@ -316,9 +332,10 @@ MqttClient.prototype.subscribe = function(topic, options, callback) {
     callback = callback || noop;
   }
   try {
+    var qos = getQoS(options && options.qos);
     var buf = this._handle._getSubscribe(topic, {
-      id: this._msgId++,
-      qos: (options && options.qos) || 0,
+      id: qos === 0 ? 0 : this._nextMsgId(),
```
Comment (yorkie) [inline src/js/mqtt.js]: Subscribe seems no need to check qos.

File: src/js/mqtt.js [hunk 14]
```diff
@@ -283,13 +297,15 @@ MqttClient.prototype.disconnect = function(err) {
...
+
   if (!Buffer.isBuffer(payload)) {
     payload = new Buffer(payload);
   }
+  var qos = getQoS(options && options.qos);
   try {
     var buf = this._handle._getPublish(topic, {
-      id: this._msgId++,
-      qos: (options && options.qos) || 0,
+      id: qos === 0 ? 0 : this._nextMsgId(),
```
Comment (yorkie) [inline src/js/mqtt.js]: The id should be undefined(unset) if QoS is zero.

File: src/js/mqtt.js (hunk 10 above)
Comment (Txiaozhe) [reply src/js/mqtt.js]: Now we have not check the ack id, so it is not very emergency now. Next we need perfect it.

Comment (yorkie) [reply src/js/mqtt.js]: Ok, seems good to me now.

File: src/js/mqtt.js [hunk 15]
```diff
@@ -274,6 +276,18 @@ MqttClient.prototype.disconnect = function(err) {
   this._socket.end();
 };
 
+MqttClient.prototype._getQoS = function(qos) {
+  return Number.isNaN(qos) ? 0 : (qos >= 0 && qos <= 2 ? qos : 0);
```
Comment (legendecas) [inline src/js/mqtt.js]: No need to check if qos is NaN, `(NaN >= 0 && NaN <= 2) === false`.

File: src/js/mqtt.js [hunk 16]
```diff
@@ -22,6 +22,8 @@ var MQTT_PINGRESP = 13;
 var MQTT_DISCONNECT = 14;
 /* eslint-enable */
 
+var MAX_MSG_ID = 0xffff;
```
Comment (yorkie) [inline src/js/mqtt.js]:
```suggestion
var MAX_PACKET_ID = 0xffff;
```

File: src/js/mqtt.js [hunk 17]
```diff
@@ -274,6 +276,18 @@ MqttClient.prototype.disconnect = function(err) {
   this._socket.end();
 };
 
+MqttClient.prototype._getQoS = function(qos) {
```
Comment (yorkie) [inline src/js/mqtt.js]:
```suggestion
MqttClient.prototype._getQoS = function getQoS(val) {
```

File: src/js/mqtt.js [hunk 18]
```diff
@@ -283,13 +297,15 @@ MqttClient.prototype.disconnect = function(err) {
...
+
   if (!Buffer.isBuffer(payload)) {
     payload = new Buffer(payload);
   }
+  var qos = this._getQoS(options && options.qos);
   try {
     var buf = this._handle._getPublish(topic, {
-      id: this._msgId++,
-      qos: (options && options.qos) || 0,
+      id: qos === 0 ? 0 : this._getNewPacketId(),
```
Comment (yorkie) [inline src/js/mqtt.js]:
```suggestion
      id: qos === 0 ? undefined : this._getNewPacketId(),
```

File: test/run_pass/test_mqtt_qos.js [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+  }, (err) => {
+    assert.strictEqual(babeee._getQoS(1), 1);
+  });
+
+  babeee.subscribe('u/love3', {
+    qos: 5
+  }, (err) => {
+    assert.strictEqual(babeee._getQoS(0), 0);
+  });
+});
```
Comment (yorkie) [inline test/run_pass/test_mqtt_qos.js]: Should assert on packetId, too.

File: src/js/mqtt.js [hunk 20]
```diff
@@ -373,7 +390,7 @@ MqttClient.prototype.reconnect = function() {
  * @method getLastMessageId
  */
 MqttClient.prototype.getLastMessageId = function() {
```
Comment (yorkie) [inline src/js/mqtt.js]: Change to `getLastPacketId`, and deprecates `getLastMessageId`?

File: src/js/mqtt.js (hunk 18 above)
Comment (Txiaozhe) [reply src/js/mqtt.js]: Can packet id be undefined?

Comment (yorkie) [reply src/js/mqtt.js]: At MQTT spec, the behavior needs to be unset if QoS is zero.

File: src/js/mqtt.js [hunk 21]
```diff
@@ -370,10 +387,10 @@ MqttClient.prototype.reconnect = function() {
 };
 
 /**
- * @method getLastMessageId
+ * @method getLastPacketId
  */
-MqttClient.prototype.getLastMessageId = function() {
-  return this._msgId;
+MqttClient.prototype.getLastPacketId = function() {
```
Comment (yorkie) [inline src/js/mqtt.js]: Backward-compatibility is required.

**Comment by yorkie** [comment]
@Txiaozhe You should add your tests to https://github.com/yodaos-project/ShadowNode/blob/master/test/testsets.json.

File: src/js/mqtt.js [hunk 22]
```diff
@@ -369,12 +386,21 @@ MqttClient.prototype.reconnect = function() {
...
+
+/**
+ * for Backward-compatibility
  * @method getLastMessageId
  */
-MqttClient.prototype.getLastMessageId = function() {
-  return this._msgId;
-};
+MqttClient.prototype.getLastMessageId = util.deprecate(_getLastPacketId, 
+  'getLastMessageId is deprecated, please remove this call.');
```
Comment (yorkie) [inline src/js/mqtt.js]:
```suggestion
  'getLastMessageId is deprecated, please use `getLastPacketId` instead.');
```

File: src/js/mqtt.js [hunk 23]
```diff
@@ -369,12 +386,21 @@ MqttClient.prototype.reconnect = function() {
...
+};
+
 /**
+ * @method getLastPacketId
+ */
+MqttClient.prototype.getLastPacketId = _getLastPacketId;
+
+/**
+ * for Backward-compatibility
  * @method getLastMessageId
```
Comment (yorkie) [inline src/js/mqtt.js]: add @deprecated is good.

File: src/js/mqtt.js (hunk 18 above)
Comment (Txiaozhe) [reply src/js/mqtt.js]:
```c
if (qos > 0)
	writeInt(&ptr, packetid);
```
if qos == 0, it will not set the packetid, so 0 is suitable.
